### PR TITLE
drivers: can: stm32 can driver rework/fixing and testcases

### PR DESCRIPTION
**Pull request by alexanderwachter**
This PR adds testcases for the can API and stm32 can driver.
Some test failed so a partial rework of the driver was done.

### CODE REVIEW

File: tests/drivers/can/api/src/main.c [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+	zassert_equal(ret, 0, "Can't set loopback-mode. Err: %d", ret);
+}
+
+/* 
+ * Sending a message to the wild should work because we are in loopback mode
+ * and therfor ACK the frame ourselves
+ */
+static void test_send_and_forget(void)
+{
+
```
Comment (jukkar) [inline tests/drivers/can/api/src/main.c]: extra empty line

File: tests/drivers/can/api/src/main.c [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+	send_test_msg(can_dev, &test_std_msg);
+}
+
+/*
+ * Test very basic filter attachment
+ * Test each type but ony one filter at a time
+ */
+static void test_filter_attach(void)
+{
+
```
Comment (jukkar) [inline tests/drivers/can/api/src/main.c]: ditto

**Comment by alexanderwachter** [comment]
@carlescufi, @nashif  Can we get this to 14.1?
The testcases unmasked some bugs in the can driver implementations.

File: tests/drivers/can/api/testcase.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+tests:
+  peripheral.can:
+    tags: driver
```
Comment (jukkar) [inline tests/drivers/can/api/testcase.yaml]: you could add "can" tag here too so that it is easier to run just the can stuff with "sanitycheck -t can"

**Review by jukkar — COMMENTED** [review]
Looks good, couple of nit-picks

File: tests/drivers/can/api/testcase.yaml (hunk 3 above)
Comment (alexanderwachter) [reply tests/drivers/can/api/testcase.yaml]: Sure.

**Comment by alexanderwachter** [comment]
Seems i did something wrong with checkpatch. I'll fit the issues in a minute

**Comment by alexanderwachter** [comment]
@lowlander this PR might be interesting for you since you'r extending this driver to multi CAN instances.

**Comment by alexanderwachter** [comment]
Just updated the userspace handlers.

**Comment by lowlander** [comment]
> @lowlander this PR might be interesting for you since you'r extending this driver to multi CAN instances.

@alexanderwachter I'll look into it in the weekend.

**Review by laperie — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**